File: android/app/src/main/java/io/frog/tasks/FrogWidgetStore.java
package io.frog.tasks;

import android.content.Context;
import android.content.SharedPreferences;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.UUID;

/**
 * The only thing the native side and the webview share.
 *
 * The webview writes a rendered snapshot of today (state) whenever app state
 * changes; the widget reads it. Actions taken on the home screen while the
 * webview is not running go the other way, into a queue the webview drains and
 * replays through the normal task endpoints. Nothing here talks to the network
 * — the native side has no session cookie and no idea who is signed in.
 */
public final class FrogWidgetStore {

    private static final String PREFS = "frog_widget";
    private static final String KEY_STATE = "state";
    private static final String KEY_QUEUE = "queue";
    private static final int MAX_QUEUE = 50;

    private static final Object LOCK = new Object();

    private FrogWidgetStore() {}

    private static SharedPreferences prefs(Context context) {
        return context.getApplicationContext()
                .getSharedPreferences(PREFS, Context.MODE_PRIVATE);
    }

    // --- state (webview -> widget) ---------------------------------------

    public static void setState(Context context, String json) {
        prefs(context).edit().putString(KEY_STATE, json).apply();
    }

    public static void clearState(Context context) {
        prefs(context).edit().remove(KEY_STATE).apply();
    }

    public static String rawState(Context context) {
        return prefs(context).getString(KEY_STATE, null);
    }

    public static JSONObject state(Context context) {
        String raw = rawState(context);
        if (raw == null) return null;
        try {
            return new JSONObject(raw);
        } catch (JSONException e) {
            return null;
        }
    }

    /** Marks a row the webview has not seen yet; it cannot be ticked. */
    public static final String PENDING_PREFIX = "pending:";

    /**
     * Shows a just-captured task straight away, before the webview has had a
     * chance to create it. The id is local-only, so the row renders as pending
     * until the real snapshot replaces it.
     */
    public static void applyLocalAdd(Context context, String text) {
        synchronized (LOCK) {
            JSONObject state = state(context);
            if (state == null) return;
            JSONArray tasks = state.optJSONArray("tasks");
            if (tasks == null) tasks = new JSONArray();
            try {
                JSONObject row = new JSONObject();
                row.put("id", PENDING_PREFIX + UUID.randomUUID());
                row.put("text", text);
                row.put("done", false);
                JSONArray next = new JSONArray();
                next.put(row);
                for (int i = 0; i < tasks.length(); i++) {
                    next.put(tasks.get(i));
                }
                state.put("tasks", next);
                state.put("totalCount", state.optInt("totalCount", 0) + 1);
                setState(context, state.toString());
            } catch (JSONException ignored) {
                // Values are all primitives; cannot throw in practice.
            }
        }
    }

    /** Flips a row locally so the widget reacts instantly, before any sync. */
    public static void applyLocalToggle(Context context, String taskId, boolean done) {
        synchronized (LOCK) {
            JSONObject state = state(context);
            if (state == null) return;
            JSONArray tasks = state.optJSONArray("tasks");
            if (tasks == null) return;
            int doneCount = 0;
            for (int i = 0; i < tasks.length(); i++) {
                JSONObject task = tasks.optJSONObject(i);
                if (task == null) continue;
                if (taskId.equals(task.optString("id"))) {
                    try {
                        task.put("done", done);
                    } catch (JSONException ignored) {
                        // Not reachable for a boolean value.
                    }
                }
                if (task.optBoolean("done")) doneCount++;
            }
            try {
                state.put("doneCount", doneCount);
            } catch (JSONException ignored) {
                // Not reachable for an int value.
            }
            setState(context, state.toString());
        }
    }

    // --- queue (widget -> webview) ---------------------------------------

    private static String currentUid(Context context) {
        JSONObject state = state(context);
        return state == null ? "" : state.optString("uid", "");
    }

    private static boolean currentGuest(Context context) {
        JSONObject state = state(context);
        return state != null && state.optBoolean("guest", false);
    }

    private static void enqueue(Context context, JSONObject action) {
        synchronized (LOCK) {
            SharedPreferences p = prefs(context);
            JSONArray queue;
            try {
                queue = new JSONArray(p.getString(KEY_QUEUE, "[]"));
            } catch (JSONException e) {
                queue = new JSONArray();
            }
            queue.put(action);
            // Drop the oldest rather than growing without bound if the app is
            // never opened again.
            while (queue.length() > MAX_QUEUE) {
                queue.remove(0);
            }
            p.edit().putString(KEY_QUEUE, queue.toString()).apply();
        }
    }

    public static void queueAdd(Context context, String text) {
        try {
            JSONObject action = new JSONObject();
            action.put("kind", "add");
            action.put("clientId", UUID.randomUUID().toString());
            action.put("text", text);
            action.put("uid", currentUid(context));
            action.put("guest", currentGuest(context));
            action.put("at", System.currentTimeMillis());
            enqueue(context, action);
        } catch (JSONException ignored) {
            // Only string/boolean/long values; cannot throw in practice.
        }
    }

    public static void queueToggle(Context context, String taskId, boolean done) {
        try {
            JSONObject action = new JSONObject();
            action.put("kind", "toggle");
            action.put("clientId", UUID.randomUUID().toString());
            action.put("taskId", taskId);
            action.put("done", done);
            action.put("uid", currentUid(context));
            action.put("guest", currentGuest(context));
            action.put("at", System.currentTimeMillis());
            enqueue(context, action);
        } catch (JSONException ignored) {
            // As above.
        }
    }

    /** Reads and empties the queue in one step, so nothing replays twice. */
    public static String drainQueue(Context context) {
        synchronized (LOCK) {
            SharedPreferences p = prefs(context);
            String queue = p.getString(KEY_QUEUE, "[]");
            p.edit().remove(KEY_QUEUE).apply();
            return queue;
        }
    }
}

File: ios/App/LiveActivities/FrogWidgetStore.swift
import Foundation
import WidgetKit

/// The only thing the app and the widget extension share.
///
/// The webview writes a rendered snapshot of today; the widget reads it. Taps
/// taken on the home screen go the other way, into a queue the webview drains
/// and replays through the normal task endpoints. Nothing here touches the
/// network — the extension has no session cookie and no idea who is signed in.
enum FrogWidgetStore {

    static let appGroup = "group.io.frog.tasks.liveactivities"
    static let kind = "FrogTasksWidget"

    private static let stateKey = "widget_state"
    private static let queueKey = "widget_queue"
    private static let maxQueue = 50
    private static let lock = NSLock()

    static var defaults: UserDefaults? {
        UserDefaults(suiteName: appGroup)
    }

    // MARK: - State (app -> widget)

    static func setState(_ json: String) {
        defaults?.set(json, forKey: stateKey)
        reload()
    }

    static func clearState() {
        defaults?.removeObject(forKey: stateKey)
        reload()
    }

    static func readState() -> WidgetState? {
        guard
            let json = defaults?.string(forKey: stateKey),
            let data = json.data(using: .utf8)
        else { return nil }
        return try? JSONDecoder().decode(WidgetState.self, from: data)
    }

    static func reload() {
        WidgetCenter.shared.reloadTimelines(ofKind: kind)
    }

    /// Flips a row locally so the widget reacts on the next redraw, before any
    /// sync. The webview's next snapshot overwrites this.
    static func applyLocalToggle(taskId: String, done: Bool) {
        lock.lock()
        defer { lock.unlock() }
        guard
            var state = readState(),
            let index = state.tasks.firstIndex(where: { $0.id == taskId })
        else { return }
        state.tasks[index].done = done
        state.doneCount = state.tasks.filter(\.done).count
        guard
            let data = try? JSONEncoder().encode(state),
            let json = String(data: data, encoding: .utf8)
        else { return }
        defaults?.set(json, forKey: stateKey)
    }

    // MARK: - Queue (widget -> app)

    static func queueToggle(taskId: String, done: Bool) {
        let state = readState()
        enqueue(PendingAction(
            kind: "toggle",
            clientId: UUID().uuidString,
            taskId: taskId,
            text: nil,
            done: done,
            uid: state?.uid ?? "",
            guest: state?.guest ?? false,
            at: Date().timeIntervalSince1970 * 1000
        ))
    }

    /// The add button. The extension can't present the composer itself, so it
    /// records the request and lets the webview open its own quick-add sheet on
    /// the next launch — the app's own route, not a URL, which is what keeps
    /// this off the address bar entirely.
    static func queueQuickAdd() {
        let state = readState()
        enqueue(PendingAction(
            kind: "quickadd",
            clientId: UUID().uuidString,
            taskId: nil,
            text: nil,
            done: nil,
            uid: state?.uid ?? "",
            guest: state?.guest ?? false,
            at: Date().timeIntervalSince1970 * 1000
        ))
    }

    private static func enqueue(_ action: PendingAction) {
        lock.lock()
        defer { lock.unlock() }
        var queue = readQueue()
        queue.append(action)
        if queue.count > maxQueue {
            queue.removeFirst(queue.count - maxQueue)
        }
        writeQueue(queue)
    }

    /// Reads and empties the queue in one step, so nothing replays twice.
    static func drainQueue() -> String {
        lock.lock()
        defer { lock.unlock() }
        let json = defaults?.string(forKey: queueKey) ?? "[]"
        defaults?.removeObject(forKey: queueKey)
        return json
    }

    private static func readQueue() -> [PendingAction] {
        guard
            let json = defaults?.string(forKey: queueKey),
            let data = json.data(using: .utf8),
            let queue = try? JSONDecoder().decode([PendingAction].self, from: data)
        else { return [] }
        return queue
    }

    private static func writeQueue(_ queue: [PendingAction]) {
        guard
            let data = try? JSONEncoder().encode(queue),
            let json = String(data: data, encoding: .utf8)
        else { return }
        defaults?.set(json, forKey: queueKey)
    }
}

// MARK: - Wire format (mirrors src/lib/widget/types.ts)

struct WidgetTask: Codable, Identifiable, Hashable {
    let id: String
    let text: String
    var done: Bool
}

struct WidgetWord: Codable, Hashable {
    let term: String
    let meaning: String
}

struct WidgetState: Codable {
    let v: Int
    let uid: String
    let guest: Bool
    let signedIn: Bool
    let day: String
    var doneCount: Int
    let totalCount: Int
    /// Which illustration medium and large draw today, picked webview-side.
    let art: String
    let word: WidgetWord
    var tasks: [WidgetTask]
    let updatedAt: Double
}

struct PendingAction: Codable {
    let kind: String
    let clientId: String
    let taskId: String?
    let text: String?
    let done: Bool?
    let uid: String
    let guest: Bool
    let at: Double
}
